test/Table_Tests/src/Common_Table_Operations/Add_Group_Number_Spec.enso

```diff
@@ -0,0 +1,131 @@
+
+main filter=Nothing = run_default_backend add_specs filter
+
+add_specs suite_builder setup =
+    if setup.is_database.not then (add_group_number_specs suite_builder setup) else
+        suite_builder.group setup.prefix+"Table.add_group_number" group_builder->
+            group_builder.specify "add_group_number should report unsupported" <|
+                table_builder = setup.light_table_builder
+                t = table_builder [['x', [1, 2, 3, 4, 5]], ['y', [5, 4, 3, 2, 1]], ['z', [1, 5, 4, 2, 3]]]
+                t.add_group_number (..Equal_Count 3) "g" . should_fail_with (Unsupported_Database_Operation.Error "add_group_number")
+
+add_group_number_specs suite_builder setup =
+    prefix = setup.prefix
+    # materialize = setup.materialize
+    # create_connection_fn = setup.create_connection_func
+
+    suite_builder.group prefix+"Table.add_group_number (common)" group_builder->
+        table_builder = setup.table_builder
+        table_builder_from_rows column_names rows = table_builder (column_names.zip rows.transpose c-> col-> [c, col])
+
+        group_builder.specify "should add group number by unique values" <|
+            t = table_builder_from_rows ['x', 'y', 'z'] [[1, 0, 2], [0, 1, 0], [1, 2, 0], [0, 1, 1], [1, 0, 1], [1, 2, 1]]
+
+            g0 = t.add_group_number (..Unique on=['x', 'y']) "g"
+            g0.at 'g' . to_vector . should_equal [0, 1, 2, 1, 0, 2]
+
+            g1 = t.add_group_number (..Unique on=['x', 'z']) "g"
+            g1.at 'g' . to_vector . should_equal [0, 1, 2, 3, 4, 4]
+
+            g2 = t.add_group_number (..Unique on=['y', 'z']) "g"
+            g2.at 'g' . to_vector . should_equal [0, 1, 2, 3, 4, 5]
+
+        group_builder.specify "should add group number by equal counts" <|
+            t = table_builder [['x', [1, 2, 3, 4, 5]], ['y', [5, 4, 3, 2, 1]], ['z', [1, 5, 4, 2, 3]]]
+
+            g0 = t.add_group_number (..Equal_Count 3) "g"
+            g0.at 'g' . to_vector . should_equal [0, 0, 1, 1, 2]
+
+            g1 = t.add_group_number (..Equal_Count 3 order_by=['x']) "g"
+            g1.at 'g' . to_vector . should_equal [0, 0, 1, 1, 2]
+
+            g2 = t.add_group_number (..Equal_Count 3 order_by=['y']) "g"
+            g2.at 'g' . to_vector . should_equal [2, 1, 1, 0, 0]
+
+            g3 = t.add_group_number (..Equal_Count 3 order_by='z') "g"
+            g3.at 'g' . to_vector . should_equal [0, 2, 1, 0, 1]
+
+            g4 = t.add_group_number (..Equal_Count 2) "g"
+            g4.at 'g' . to_vector . should_equal [0, 0, 0, 1, 1]
+
+            g5 = t.add_group_number (..Equal_Count 2 order_by=['x']) "g"
+            g5.at 'g' . to_vector . should_equal [0, 0, 0, 1, 1]
+
+            g6 = t.add_group_number (..Equal_Count 2 order_by=['y']) "g"
+            g6.at 'g' . to_vector . should_equal [1, 1, 0, 0, 0]
+
+            g7 = t.add_group_number (..Equal_Count 2 order_by='z') "g"
+            g7.at 'g' . to_vector . should_equal [0, 1, 1, 0, 0]
```

Comment:
Would be good to also test edge cases:
- `..Equal_Count 0` -> should fail with illegal argument I imagine
- `..Equal_Count -1` -> also should fail, cannot be negative
- `..Equal_Count 1` -> should work and just assign the same group number to all entries - not very useful on its own, but may be useful when the number of groups is a parameter coming from somewhere - splitting into 1 group is technically a valid thing to do

Reply:
Done